feat: [#504] Route supports configure timeout

KlassnayaAfrodita opened the pull request:
https://github.com/goravel/goravel/issues/504

add withcontext

## 📑 Description

Closes https://github.com/goravel/goravel/issues/504

<!-- Please add Review Ready tag when the PR is good to go -->
<!-- More description can be written after this -->

<!-- Do not remove this line -->
@coderabbitai summary
<!-- Trigger AI description by commenting @coderabbitai summary -->

## ✅ Checks

<!-- Make sure your PR passes the CI checks and do check the following fields as needed - -->
- [ ] Added test cases for my code

KlassnayaAfrodita committed:
Update context.go

https://github.com/goravel/goravel/issues/504

add withcontext

hwbrzzl reviewed (commented):
This new method is good, please add the global configuration logic.

hwbrzzl commented:
Please merge your two PRs.

KlassnayaAfrodita committed:
Update context_test.go

https://github.com/goravel/gin/pull/97

KlassnayaAfrodita commented:
I merged two PCs. Please explain what "add the global configuration logic" means

hwbrzzl commented:
<img width="923" alt="image" src="https://github.com/user-attachments/assets/fba5a909-685e-4bcc-8865-869e17615ac9">

KlassnayaAfrodita commented:
I added a timeout to the config, how to use them in gin and fiber?

KlassnayaAfrodita commented:
Do I understand correctly what needs to be added to goravel/gin/route.go in the function NewRoute(config config.Config, parameters map[string]any) (*Route, error) add 
// ...
    router.Use(gin.WithTimeout(config.Timeout))
// ...
?

KlassnayaAfrodita committed:
Update route.go

hwbrzzl commented:
CI failed.

hwbrzzl commented:
gin doesn't have the `WithTimeout` method.

KlassnayaAfrodita commented:
do I need to implement TimeoutMiddleware?

hwbrzzl commented:
Yes, don't use gin-contrib/timeout, it has a lot of dependencies, we can implement it ourselves.

KlassnayaAfrodita commented:
> Yes, don't use gin-contrib/timeout, it has a lot of dependencies, we can implement it ourselves.

where to place middleware timeout?

hwbrzzl commented:
Follow: https://github.com/goravel/gin/blob/master/tls.go

KlassnayaAfrodita committed:
Create timeoutMiddleware.go

https://github.com/goravel/gin/pull/97

KlassnayaAfrodita committed:
Update route.go

update NewRoute 
https://github.com/goravel/gin/pull/97

KlassnayaAfrodita commented:
I added timeout middleware following https://github.com/goravel/gin/blob/master/tls.go and rewrote the use of this middleware in NewRoute. Please check my changes and give feedback

KlassnayaAfrodita committed:
Update timeoutMiddleware.go

KlassnayaAfrodita committed:
Update timeoutMiddleware.go

KlassnayaAfrodita committed:
Update route.go

KlassnayaAfrodita committed:
Update route.go

KlassnayaAfrodita committed:
Update timeoutMiddleware.go

hwbrzzl commented on timeoutMiddleware.go:
```diff
@@ -0,0 +1,40 @@
+		// Refreshing the request context with a timeout
+		ctx.WithContext(timeoutCtx)
+
+		// We start executing the request in a new goroutine
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			ctx.Request().Next()
+		}()
+
+		select {
+		case <-done:
+			// The request completed before the timeout expired
+		case <-timeoutCtx.Done():
+			// After the timeout expires, return the status 504 Gateway Timeout
+			if timeoutCtx.Err() == context.DeadlineExceeded {
+				ctx.Response().Writer().WriteHeader(http.StatusGatewayTimeout)
+				_, _ = ctx.Response().Writer().Write([]byte("Request timed out"))
+			}
+		}
```
I doubt this's unnecessary, when we set `timeCtx` to `gin`, `gin` will handle this automatically. We can test it.

KlassnayaAfrodita replied:
I'll trust your commit, where can I read about what you say?

KlassnayaAfrodita replied:
I tested on localhost, gin does not cancel the request when WithTimeout expires. for example:
package main

import (
	"context"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

func main() {
	router := gin.Default()

	router.Use(func(c *gin.Context) {
		timeoutCtx, cancel := context.WithTimeout(c.Request.Context(), 3*time.Second)
		defer cancel()
		c.Request = c.Request.WithContext(timeoutCtx)
		c.Next()
	})

	router.GET("/long_request", func(c *gin.Context) {
		time.Sleep(6 * time.Second)
		c.String(http.StatusOK, "This should time out")
	})

	router.Run(":8080")
}

curl http://localhost:8080/long_request

the server continues processing the request after the context is canceled

hwbrzzl replied:
Good catch 👍

hwbrzzl commented on timeoutMiddleware.go:
```diff
@@ -0,0 +1,40 @@
+// TimeoutMiddleware creates middleware to set a timeout for a request
+func TimeoutMiddleware() contractshttp.Middleware {
+	return func(ctx contractshttp.Context) {
+		timeout := config.GetInt("http.http.drivers.gin.timeout", 3) * time.Second
```
```suggestion
		timeout := config.GetInt("http.request_timeout", 3) * time.Second
```

KlassnayaAfrodita commented:
what else should I do about this issue?

hwbrzzl commented on timeoutMiddleware.go:
```diff
@@ -0,0 +1,40 @@
+		ctx.WithContext(timeoutCtx)
+
+		// We start executing the request in a new goroutine
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			ctx.Request().Next()
+		}()
```
Could you try the logic below? Not sure if the routine is required.

```suggestion
		ctx.Request().Next()
```

hwbrzzl commented on timeoutMiddleware.go:
```diff
@@ -0,0 +1,40 @@
+		}()
+
+		select {
+		case <-done:
```
And modify this with above simultaneously.

```suggestion
		case <-ctx.Request().Origin().Context().Done():
```

KlassnayaAfrodita replied:
of course, so I need to remove the select block?

hwbrzzl replied:
No, as you tested, only setting context is not enough, `select` is required. I mean changing `done` to `ctx.Request().Origin().Context().Done()`, test it, check if it's correct.

KlassnayaAfrodita replied:
I tested it, it worked, I'll make changes today

hwbrzzl commented on timeoutMiddleware.go:
The file name can be `timeout_middleware.go`, and change `tls.go` to `tls_middleware.go`.

KlassnayaAfrodita committed:
Update timeoutMiddleware.go

KlassnayaAfrodita committed:
Rename timeoutMiddleware.go to timeout_middleware.go

KlassnayaAfrodita committed:
Rename tls.go to tls_middleware.go

KlassnayaAfrodita committed:
Update timeout_middleware.go

KlassnayaAfrodita committed:
Update timeout_middleware.go

hwbrzzl commented on timeout_middleware.go:
```diff
@@ -0,0 +1,32 @@
+		case <-ctx.Request().Origin().Context().Done():
+			
+			if timeoutCtx.Err() == context.DeadlineExceeded {
+				ctx.Response().Writer().WriteHeader(http.StatusGatewayTimeout)
+				_, _ = ctx.Response().Writer().Write([]byte("Request timed out"))
```
```suggestion
				ctx.Request().AbortWithStatus(http.StatusGatewayTimeout)
```

hwbrzzl replied:
Please update this.

hwbrzzl commented on route.go:
```diff
@@ -77,7 +81,7 @@ func (r *Route) Fallback(handler httpcontract.HandlerFunc) {
 
 func (r *Route) GlobalMiddleware(middlewares ...httpcontract.Middleware) {
-	middlewares = append(middlewares, Cors(), Tls())
+	middlewares = append(middlewares, Cors(), Tls(), TImeoutMiddleware())
```
```suggestion
	middlewares = append(middlewares, Cors(), Tls(), TimeoutMiddleware())
```

hwbrzzl replied:
Please update this.

hwbrzzl commented on route.go:
```diff
@@ -30,6 +30,10 @@ func NewRoute(config config.Config, parameters map[string]any) (*Route, error) {
 	engine := gin.New()
 	engine.MaxMultipartMemory = int64(config.GetInt("http.drivers.gin.body_limit", 4096)) << 10
 	engine.Use(gin.Recovery()) // recovery middleware
+	// timeoutMiddleware
+	// timeout := TimeoutMiddleware(config.GetInt("http.request_timeout", 3) * time.Second)
+	// engine.Use(middlewaresToGinHandlers([]httpcontract.Middleware{timeout})...)
+	
```
Remove it.

hwbrzzl commented on route.go:
Please add test cases for the new logic in `route_test.go`.

KlassnayaAfrodita committed:
Update timeout_middleware.go

KlassnayaAfrodita committed:
Update route.go

KlassnayaAfrodita committed:
Update route_test.go

KlassnayaAfrodita committed:
Update route_test.go

KlassnayaAfrodita committed:
Update route_test.go

KlassnayaAfrodita committed:
Update route_test.go

KlassnayaAfrodita commented:
please check my changes

hwbrzzl reviewed (commented):
Okay, according to the change in this PR, I think it's time to merge https://github.com/goravel/framework/pull/675, please fix the CI.

After the PR above is merged, please execute `go get github.com/goravel/framework@master` in this PR, to make sure the CI of this PR passes.

KlassnayaAfrodita commented:
https://github.com/goravel/gin/pull/97#discussion_r1807611415 https://github.com/goravel/gin/pull/97#discussion_r1807611511 already implemented

KlassnayaAfrodita committed:
Update timeout_middleware.go

KlassnayaAfrodita committed:
Update route.go

KlassnayaAfrodita committed:
Update route_test.go

KlassnayaAfrodita commented:
Please, check route_test.go. I think TimeoutMiddleware is not being called

KlassnayaAfrodita committed:
Update route_test.go

KlassnayaAfrodita committed:
Update timeout_middleware.go

KlassnayaAfrodita commented:
with this version, everything works for me

KlassnayaAfrodita committed:
Update timeout_middleware.go

KlassnayaAfrodita committed:
Update route_test.go

KlassnayaAfrodita commented:
I constantly get a null pointer to the request, that is, when ctx.Request() is called.Next() gives an error that the pointer is null

KlassnayaAfrodita commented:
I constantly get a null pointer to the request, that is, when ctx.Request() is called.Next() gives an error that the pointer is null

hwbrzzl committed:
optimize timeout middleware

hwbrzzl commented:
Could you add me to this repo? https://github.com/KlassnayaAfrodita/gin
I optimized your code and want to push my code to your repo, so that you can check the optimization.

KlassnayaAfrodita commented:
> hwbrzzl

yes, I invited you. patch-1 current branch

KlassnayaAfrodita committed:
Merge pull request #1 from KlassnayaAfrodita/bowen/timeout-middleware

feat: optimize timeout middleware

hwbrzzl commented:
Please solve the conflict.

KlassnayaAfrodita committed:
Merge branch 'master' into patch-1

hwbrzzl commented on middleware_timeout_test.go:
```diff
@@ -0,0 +1,46 @@
+	require.NoError(t, err)
+
+	route.Middleware(Timeout(1*time.Second)).Get("/timeout", func(ctx contractshttp.Context) contractshttp.Response {
+		time.Sleep(1 * time.Second)
```
The test case failed in Windows: https://github.com/goravel/gin/actions/runs/11534959849/job/32113680328?pr=97

We may try to change this to 2 second:

```suggestion
		time.Sleep(2 * time.Second)
```

<img width="770" alt="image" src="https://github.com/user-attachments/assets/25debba4-5ed4-4aa4-9dfc-f2165746fc2e">

KlassnayaAfrodita committed:
Update middleware_timeout_test.go

hwbrzzl reviewed (approved):
Great  PR, thanks!

hwbrzzl commented:
We can optimize fiber next.
